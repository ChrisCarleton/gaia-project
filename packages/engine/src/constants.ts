import {
  FactionType,
  PlanetType,
  ResearchArea,
  StructureType,
} from './interfaces';

export const ColonizingStructures: Set<StructureType> = new Set([
  StructureType.Academy,
  StructureType.Mine,
  StructureType.PlanetaryInstitute,
  StructureType.ResearchLab,
  StructureType.TradingStation,
]);

export const FactionTypeNames: Record<FactionType, string> = {
  [FactionType.Ambas]: 'Ambas',
  [FactionType.BalTaks]: "Bal T'Aks",
  [FactionType.Bescods]: 'Bescods',
  [FactionType.Firaks]: 'Firaks',
  [FactionType.Geodens]: 'Geodens',
  [FactionType.Gleens]: 'Gleens',
  [FactionType.HadschHallas]: 'Hadsch Hallas',
  [FactionType.Itars]: 'Itars',
  [FactionType.Ivits]: 'Ivits',
  [FactionType.Lantids]: 'Lantids',
  [FactionType.Nevlas]: 'Nevlas',
  [FactionType.Taklons]: 'Taklons',
  [FactionType.Terrans]: 'Terrans',
  [FactionType.Xenos]: 'Xenos',
} as const;

export const PlanetTypeNames: Record<PlanetType, string> = {
  [PlanetType.Desert]: 'Desert',
  [PlanetType.Gaia]: 'Gaia',
  [PlanetType.Ice]: 'Ice',
  [PlanetType.Oxide]: 'Oxide',
  [PlanetType.Swamp]: 'Swamp',
  [PlanetType.Terra]: 'Earth-like',
  [PlanetType.Titanium]: 'Titanium',
  [PlanetType.Transdim]: 'Transdim',
  [PlanetType.Volcanic]: 'Volcanic',
} as const;

export const ResearchAreaNames: Record<ResearchArea, string> = {
  [ResearchArea.AI]: 'Artificial Intelligence',
  [ResearchArea.Economics]: 'Economy',
  [ResearchArea.Gaia]: 'Gaia Project',
  [ResearchArea.Navigation]: 'Navigation',
  [ResearchArea.Science]: 'Science',
  [ResearchArea.Terraforming]: 'Terraforming',
} as const;

export const StructureTypeNamesPlural: Record<StructureType, string> = {
  [StructureType.Academy]: 'Acadamies',
  [StructureType.Gaiaformer]: 'Gaiaformers',
  [StructureType.Mine]: 'Mines',
  [StructureType.PlanetaryInstitute]: 'Planetary institute', // There is no pluralization for this... players only get one.
  [StructureType.ResearchLab]: 'Research labs',
  [StructureType.TradingStation]: 'Trading stations',
};
