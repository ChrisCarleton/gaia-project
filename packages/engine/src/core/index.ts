export * from './terraforming-costs';
export * from './config';
export * from './game-context';
export * from './game';

export * from '../maps/basic-map-model';
