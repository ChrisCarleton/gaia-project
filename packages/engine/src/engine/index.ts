export * from './errors';

export * from './lobby';
