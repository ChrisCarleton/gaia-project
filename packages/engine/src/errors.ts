export enum ErrorCode {
  // Player selection.
  TooFewPlayers,
  TooManyPlayers,
  FactionConflict,

  // State related errors
  ActionNotSupportedNow,
  GameAlreadyBegun,
  GameNotStarted,

  // Invalid action errors
  InvalidMinePlacement,
  InvalidRoundBooster,
  ResearchTrackTierFiveOccupied,
  RoundBoosterNotSelected,

  // Resource issues
  InsufficientKnowledge,
  InsufficientPower,
}

export class GPError extends Error {
  constructor(
    readonly code: ErrorCode,
    message: string,
  ) {
    super(message);
    Object.setPrototypeOf(this, GPError.prototype);
  }
}
