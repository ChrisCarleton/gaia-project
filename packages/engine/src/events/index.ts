export * from './event-args';
export * from './observer';
