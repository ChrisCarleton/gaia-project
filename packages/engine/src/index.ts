export * from './errors';
export * from './interfaces';

export * from './events';
export * from './core';
export * from './factions';
export * from './players';
