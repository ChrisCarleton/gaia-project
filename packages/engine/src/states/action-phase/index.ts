export * from './free-actions';
export * from './pass-action';
export * from './research-action';

export * from './determine-next-state';
