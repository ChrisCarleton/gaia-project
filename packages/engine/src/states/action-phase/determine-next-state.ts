import { ObserverPublisher } from '../../events';
import {
  ChangeStateFunction,
  GameContext,
  GameState,
  State,
} from '../../interfaces';
import { loadState } from '../load-state';

export class NextStateDetermination {
  static determineNextState(
    context: GameContext,
    events: ObserverPublisher,
    changeState: ChangeStateFunction,
  ): State {
    const { currentRound, currentPlayer: player, players } = context;

    // Find the current player in turn order.
    let index = players.findIndex((p) => Object.is(p, player));
    if (index === -1) {
      throw new Error(
        'Unexpected error. Could not find player in game context.',
      );
    }

    // Find the next player in turn order that has not already passed.
    do {
      index = (index + 1) % players.length;
      if (!players[index].passed) {
        return loadState(
          {
            type: GameState.ActionPhase,
            player: index,
          },
          context,
          events,
          changeState,
        );
      }
    } while (!Object.is(player, players[index]));

    // If everyone has passed then we can clean up for the next round, or end the game if
    // we are on the final round.
    return currentRound < 6
      ? loadState({ type: GameState.IncomePhase }, context, events, changeState)
      : loadState({ type: GameState.GameEnded }, context, events, changeState);
  }
}
