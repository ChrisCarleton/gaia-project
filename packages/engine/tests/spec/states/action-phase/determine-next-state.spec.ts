import { ObserverPublisher } from '../../../../src';
import { NextStateDetermination } from '../../../../src/states/action-phase/determine-next-state';
import {
  TestObserver,
  createTestContext,
  createTestPlayer,
} from '../../../util';

describe('Next State Determination For Action Phase', () => {
  let events: ObserverPublisher;

  beforeEach(() => {
    events = new TestObserver();
  });

  it('will advance to the next player in turn order (first player)', () => {
    const players = [
      createTestPlayer(),
      createTestPlayer(),
      createTestPlayer(),
      createTestPlayer(),
    ];
    const context = createTestContext({ players });

    const options = NextStateDetermination.determineNextState(
      context,
      events,
      jest.fn(),
    );

    expect(options).toMatchSnapshot();
  });

  it('will advance to the next player in turn order (third player)', () => {
    const players = [
      createTestPlayer(),
      createTestPlayer(),
      createTestPlayer(),
      createTestPlayer(),
    ];
    const context = createTestContext({ players, currentPlayer: players[2] });

    const options = NextStateDetermination.determineNextState(
      context,
      events,
      jest.fn(),
    );

    expect(options).toMatchSnapshot();
  });

  it('will return back to the start of turn order if we reach the end and not all players have passed', () => {
    const players = [
      createTestPlayer(),
      createTestPlayer(),
      createTestPlayer(),
      createTestPlayer(),
    ];
    const context = createTestContext({ players, currentPlayer: players[3] });

    const options = NextStateDetermination.determineNextState(
      context,
      events,
      jest.fn(),
    );

    expect(options).toMatchSnapshot();
  });

  it('will skip over players who have already previously passed', () => {
    const players = [
      createTestPlayer(),
      createTestPlayer({ passed: true }),
      createTestPlayer({ passed: true }),
      createTestPlayer(),
    ];
    const context = createTestContext({ players, currentPlayer: players[0] });

    const options = NextStateDetermination.determineNextState(
      context,
      events,
      jest.fn(),
    );

    expect(options).toMatchSnapshot();
  });

  it('will skip over players who have already passed AND loop back to the beginning of turn order if necessary', () => {
    const players = [
      createTestPlayer(),
      createTestPlayer(),
      createTestPlayer({ passed: true }),
      createTestPlayer({ passed: true }),
    ];
    const context = createTestContext({ players, currentPlayer: players[1] });

    const options = NextStateDetermination.determineNextState(
      context,
      events,
      jest.fn(),
    );

    expect(options).toMatchSnapshot();
  });

  for (let round = 1; round < 6; round++) {
    it(`will advance to clean up phase if all players have passed in round ${round}`, () => {
      const players = [
        createTestPlayer({ passed: true }),
        createTestPlayer({ passed: true }),
        createTestPlayer({ passed: true }),
        createTestPlayer({ passed: true }),
      ];
      const context = createTestContext({
        players,
        currentPlayer: players[1],
        currentRound: round,
      });

      const options = NextStateDetermination.determineNextState(
        context,
        events,
        jest.fn(),
      );

      expect(options).toMatchSnapshot();
    });
  }

  it('will advance to game ended phase if all players have passed in the final round', () => {
    const players = [
      createTestPlayer({ passed: true }),
      createTestPlayer({ passed: true }),
      createTestPlayer({ passed: true }),
      createTestPlayer({ passed: true }),
    ];
    const context = createTestContext({
      players,
      currentPlayer: players[1],
      currentRound: 6,
    });

    const options = NextStateDetermination.determineNextState(
      context,
      events,
      jest.fn(),
    );

    expect(options).toMatchSnapshot();
  });
});
