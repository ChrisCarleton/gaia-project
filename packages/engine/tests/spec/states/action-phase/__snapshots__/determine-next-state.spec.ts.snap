// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Next State Determination For Action Phase will advance to clean up phase if all players have passed in round 1 1`] = `
{
  "type": "incomePhase",
}
`;

exports[`Next State Determination For Action Phase will advance to clean up phase if all players have passed in round 2 1`] = `
{
  "type": "incomePhase",
}
`;

exports[`Next State Determination For Action Phase will advance to clean up phase if all players have passed in round 3 1`] = `
{
  "type": "incomePhase",
}
`;

exports[`Next State Determination For Action Phase will advance to clean up phase if all players have passed in round 4 1`] = `
{
  "type": "incomePhase",
}
`;

exports[`Next State Determination For Action Phase will advance to clean up phase if all players have passed in round 5 1`] = `
{
  "type": "incomePhase",
}
`;

exports[`Next State Determination For Action Phase will advance to game ended phase if all players have passed in the final round 1`] = `
{
  "type": "gameEnded",
}
`;

exports[`Next State Determination For Action Phase will advance to the next player in turn order (first player) 1`] = `
{
  "player": 1,
  "type": "actionPhase",
}
`;

exports[`Next State Determination For Action Phase will advance to the next player in turn order (third player) 1`] = `
{
  "player": 3,
  "type": "actionPhase",
}
`;

exports[`Next State Determination For Action Phase will return back to the start of turn order if we reach the end and not all players have passed 1`] = `
{
  "player": 0,
  "type": "actionPhase",
}
`;

exports[`Next State Determination For Action Phase will skip over players who have already passed AND loop back to the beginning of turn order if necessary 1`] = `
{
  "player": 0,
  "type": "actionPhase",
}
`;

exports[`Next State Determination For Action Phase will skip over players who have already previously passed 1`] = `
{
  "player": 3,
  "type": "actionPhase",
}
`;
